File: ocaml/lab2/goldbach.ml
let is_prime x =
  let rec aux n =
    if n > (x/2) then
      true
    else
      if x mod n = 0 then false else aux (n+1)
  in aux 2

let rec all_primes n =
  if n = 1 then 
    [] 
  else
    if is_prime n then n :: (all_primes (n-1)) else all_primes (n-1) 
  
let goldbach_list n =
  List.map (fun x -> (x, n-x)) (
    List.filter (fun x -> is_prime (n-x)) (all_primes n)
  )

let goldbach n = 
  match goldbach_list n with
  | x::xs -> x
  | _ -> failwith "nobel per la matematica"

File: scala/lab1/Goldbach.scala
def is_prime(n: Int): Boolean = {
    def prime(n: Int, x: Int): Boolean = {
        if (x > n/2) return true;
        if (n % x == 0) return false;
        
        return prime(n, x+1);
    };
    return prime(n, 2);
}

def goldbach(n: Int): (Int, Int) = {
    def gb(x: Int): (Int, Int) = {
        if (is_prime(x) && is_prime(n-x))
            return (x, n-x);
        else return gb(x+1);
    }

    return gb(1);
}

def goldbach_list(n: Int, m: Int): List[(Int, Int)] = {
    return {for (x <- List.range(n, m); if x % 2 == 0) yield goldbach(x)};
}